File: output/常规思路，每遍历k个做一翻转/常规思路，每遍历k个做一翻转.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
//输入为待翻转链表的前一个结点，输出仍为此结点，但链表已翻转
struct ListNode* reverse(struct ListNode* head) {
    struct ListNode *curfront,*cur,*curnext;
    if (head->next == NULL) {
        return NULL;
    }

    curfront = head->next;
    cur = curfront->next;
    curfront->next = NULL;

    while (cur) {
        curnext = cur->next;
        cur->next = curfront;
        curfront = cur;
        cur = curnext;
    }
    head->next = curfront;
    return head;
}
struct ListNode* reverseKGroup(struct ListNode* head, int k){
    struct ListNode* dummyhead = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *curfront = NULL,*curnext = NULL,*cur = NULL;
    int i = 1;
    if (head == NULL) return NULL;
    dummyhead->next = head;
    curfront = dummyhead;
    cur = dummyhead->next;

    while (cur) {
        if ((cur = cur->next) == NULL) break;
        if (++i == k) {
            curnext = cur->next;
            cur->next = NULL;//之前漏掉了断开链接这一步
            cur = curfront->next;//存下翻转后的尾
            reverse(curfront);
            cur->next = curnext;//断开链接之后还要链上
            curfront = cur;
            cur = curnext;
            i = 1;
        }
    }

    return dummyhead->next;

    
 }

File: output/常规思路，每遍历k个做一翻转/常规思路，每遍历k个做一翻转.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val
        }
    }
}

fn reverse(head: &mut Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    if let Some(ref mut head_node) = head {
        if head_node.next.is_none() {
            return None;
        }
        
        let mut curfront = head_node.next.take();
        if let Some(ref mut curfront_node) = curfront {
            let mut cur = curfront_node.next.take();
            
            while let Some(mut cur_node) = cur {
                let curnext = cur_node.next.take();
                cur_node.next = curfront;
                curfront = Some(cur_node);
                cur = curnext;
            }
        }
        
        head_node.next = curfront;
        head.clone()
    } else {
        None
    }
}

fn reverse_k_group(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    if head.is_none() {
        return None;
    }
    
    let mut dummy_head = Box::new(ListNode::new(0));
    dummy_head.next = head;
    let mut dummy_head = Some(dummy_head);
    
    let mut i = 1;
    let mut curfront = &mut dummy_head;
    
    loop {
        // Count k nodes
        let mut cur = curfront.as_ref().unwrap().next.as_ref();
        let mut count = 0;
        
        while let Some(node) = cur {
            count += 1;
            if count == k {
                break;
            }
            cur = node.next.as_ref();
        }
        
        if count < k {
            break;
        }
        
        // Find the k-th node and break the connection
        let mut cur_ref = &mut curfront.as_mut().unwrap().next;
        for _ in 0..k-1 {
            cur_ref = &mut cur_ref.as_mut().unwrap().next;
        }
        
        let curnext = cur_ref.as_mut().unwrap().next.take();
        
        // Reverse the k nodes
        let tail = curfront.as_ref().unwrap().next.as_ref().unwrap().val;
        reverse(curfront);
        
        // Find the new tail (original head of the k-group)
        let mut new_tail = &mut curfront.as_mut().unwrap().next;
        while new_tail.as_ref().unwrap().next.is_some() {
            new_tail = &mut new_tail.as_mut().unwrap().next;
        }
        
        // Connect to the rest
        new_tail.as_mut().unwrap().next = curnext;
        
        // Move curfront to the new tail
        while curfront.as_ref().unwrap().next.as_ref().unwrap().next.is_some() {
            curfront = unsafe { 
                std::mem::transmute(&mut curfront.as_mut().unwrap().next)
            };
        }
    }
    
    dummy_head.unwrap().next
}